raft_test.go

```diff
@@ -1703,6 +1785,9 @@ func TestRaft_UserSnapshot(t *testing.T) {
 	conf := inmemConfig(t)
 	conf.SnapshotThreshold = 50
 	conf.TrailingLogs = 10
+	// Set protocol to a lower version to avoid applying configuration entries on
+	// bootstrap
+	conf.ProtocolVersion = 1
```

Comment:
Didn't look at the context but does this mean this test will not be covering the v2+ case of the thing it's meant to be testing any more? If so, is there a way we can make it pass reliably without limiting the coverage of the current protocol version - like have the mock able to turn off config persistence where we don't care about testing that?

Reply:
I went down this path originally but it was turning out to be a major change to the testing framework. I don't believe this test covers any differences between v1 and v2. It just tests that snapshot returns errors in certain circumstances

Reply:
🤔 Perhaps I'm missing something but is it not possible to have another field on MockStore that optionally enables the new interface and if it's not true then just make the new Implementation method a no-op?

Then all old tests would work just as before presumably since there is no state change for config updates but we could still write the new test with just a one liner to enable the ConfigStore interface?

Reply:
The issue is the test expects Raft to report a 0 Applied index on first snapshot. But if the FSM even implements the ConfigurationStore interface then the FSM will increment its applied index for the initial bootstrap call. So a no-op isn't enough, we would need two FSM types, one that implements the new interface and one that doesn't. This is where the large amount of test framework changes came in, the cluster object assumes the FSM's type to be *MockFSM a number of places

Reply:
Gotcha.

I can let it go if there is no clean way but would it work to do something like:

```go
type MockPersistentFSM struct {
  MockFSM
  configurations []Configuration
}
// Implement ConfigurationStore for MockPersistentFSM
```

Then the existing tests all get the same behaviour as before, new tests can use the new type and any helper that expects `*MockFSM` can be passed `&fsm.MockFSM`?

If it gets more complex than that I guess this is reasonable for now and is not super likely to change in the future but it feels like at least a _potential_ footgun if a future change to the code path that is protocol version dependent will seem to pass all tests fine but actually won't be exercised here.